programs/prettier.nix

```diff
@@ -0,0 +1,282 @@
+      description = "Enforce single attribute per line in HTML, Vue and JSX.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    singleQuote = mkOption {
+      description = "Use single quotes instead of double quotes.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    tabWidth = mkOption {
+      description = "Number of spaces per indentation level.";
+      type = types.nullOr types.int;
+      example = 2;
+      default = null;
+    };
+    trailingComma = mkOption {
+      description = "Print trailing commas wherever possible when multi-line.";
+      type = types.nullOr (types.enum [ "es5" "none" "all" ]);
+      example = "es5";
+      default = null;
+    };
+    useTabs = mkOption {
+      description = "Indent with tabs instead of spaces";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    vueIndentScriptAndStyle = mkOption {
+      description = "Indent script and style tags in Vue files.";
+      type = types.nullOr types.bool;
+      example = false;
+      default = null;
+    };
+    overrides = mkOption {
+      description = "Provide a list of patterns to override prettier configuration.";
+      type = types.nullOr (types.listOf types.attrs);
+      example = { files = [ "*.html" "legacy/**/*.js" ]; options.tabwidth = 4; };
+      default = null;
+    };
+  };
+
+  # We filter out null values in dprint configuration since dprint errors when
+  # they're present
+  mkConfigFile = { filename ? ".prettierrc.json", config ? { } }:
+    configFormat.generate filename (filterAttrsRecursive (n: v: v != null) config);
+
+in
+{
+  options.programs.prettier = {
+    enable = mkEnableOption "prettier";
+    package = mkOption {
+      type = types.package;
+      default = pkgs.nodePackages.prettier;
+      defaultText = lib.literalExpression "pkgs.nodePackages.prettier";
+      description = lib.mdDoc ''
+        prettier derivation to use.
+      '';
+    };
+
+    # Represents the .prettierrc.json config schema
+    config = configSchema;
+
+    # Prettier-specific includes override
+    includes = mkOption {
+      description = "Path / file patterns to include for Prettier";
+      type = types.listOf types.str;
+      default = [
+        "*.css"
+        "*.html"
+        "*.js"
+        "*.json"
+        "*.jsx"
+        "*.md"
+        "*.mdx"
+        "*.scss"
+        "*.ts"
+        "*.tsx"
+        "*.yaml"
+        "*.yml"
+      ];
+    };
+
+    # Prettier-specific includes override
+    excludes = mkOption {
+      description = "Path / file patterns to exclude for Prettier";
+      type = types.listOf types.str;
+      default = [ ];
+    };
+
+    # Wrapped prettier invocation passing a generated .prettierrc.json
+    # configuration
+    wrapper = mkOption {
+      description = "The prettier package, wrapped with the config file.";
```

Comment:
this option should be marked as private as it's not meant to be used directly by users